[FLOC-3301] add a server providing a basic http api

avg-I opened the pull request:
Currently only submit method is implemented, so the server is not really useful.
Also, the submitted is not really persisted, a real backend for doing that is going to be added later.

avg-I committed:
add basic http api for storing the benchmarking results

At the moment the results are simply discarded, there is no real backend
to persist them.

avg-I committed:
add new requirements and move all of them to requirements.txt

avg-I committed:
add a basic start-up shell script

avg-I committed:
add more useful for testing InMemoryBackend and some tests http submit api

jml commented on benchmark/_interfaces.py:
```diff
@@ -0,0 +1,48 @@
+# Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
```
We don't put the year in Copyright lines.

avg-I replied:
I think that I picked that template from `benchmark` folder in `flocker`. But I see that the prevailing majority of the code does not use the year. So, I am fixing this.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,123 @@
+        return d
+
+
+class BenchmarkAPITest(unittest.TestCase):
```
We're moving away from using Trial TestCase in flocker, and are instead using `flocker.testtools.AsyncTestCase`.

Please either use this, making `flocker` a dependency, or use the testtools library, writing this test as:

```python
class BenchmarkAPITest(TestCase):
    """
    Tests for ``BenchmarkAPI``.
    """

    run_tests_with = AsynchronousDeferredRunTest
```

jml replied:
Normal naming is `FooTests`, so please change this to `BenchmarkAPITests`

avg-I replied:
I decided to use the `testtools` library directly. I am making the suggested changes. Additionally, I had to provide `timeout` to `AsynchronousDeferredRunTest`, because the default one of 0.005 seconds was not enough to test HTTP requests over TCP/IP.

jml replied:
Thanks :)

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,123 @@
+        """
+        Returned content is the expected JSON.
+        """
+        req = self.test_submit_success()
```
Rather than delegate to another test method, please extract the common functionality out of the test method (I think it's just `self.agent.request("POST"", "/submit", bodyProducer=StringProducer(self.RESULT_JSON))`) to a helper method that does the action (maybe making the HTTP method, URL and body data all be params?), and then use that in the tests.

Then, all of the actual assertions about how that behaves can go in the tests. 

This makes it more clear as to what each test is actually testing.

jml commented on benchmark/httpapi.py:
```diff
@@ -0,0 +1,215 @@
+        return ""
+
+    @app.route("/submit", methods=['POST'])
+    def submit(self, request):
```
Apologies if I'm covering ground that's already been covered. 

I think we would generally prefer to have a more RESTful API. That would mean having an endpoint for `/benchmark-results` (or similar) and then POSTing to that when we want to add a new one, rather than having a dedicated endpoint for submitting.

avg-I replied:
I am not sure if I got your suggestion.
Could you please take a look at https://clusterhq.atlassian.net/browse/FLOC-3302 ?
I implemented the query interface as another `POST` method, so for me it did make sense to have different URLs for the interfaces.

jml replied:
The query interface should be a `GET`, not a `POST`. 

POST:

avg-I replied:
Seems like your comment somehow got posted incomplete.

I had many doubts over the query interface.
In the end I went for POST, because I wanted to submit JSON as a filter / template.
Doing that through arguments in a GET request is not as natural as just placing JSON into a body of the POST request, despite the GET being more natural for a query in general.

I did some searching on stackoverflow and the "consensus" there seems to be that using the POST for a query is not unacceptable...

jml commented on benchmark/httpapi.py:
```diff
@@ -0,0 +1,215 @@
+            return dumps(result)
+
+        d.addCallback(stored)
+        return d
```
I strongly suggest adding a way to retrieve results from the server in this patch. Until that happens, the submit functionality isn't useful. Further, by being able to roundtrip, we'll discover new things about the design. 

e.g. This should really return a `201 Created` response code, and the response should include a `Location` header that points to the newly created resource.

avg-I replied:
I created this PR for the submit functionality only because there are two different issues for the submit and query features.  The other one being FLOC-3302. Although, I see that your suggestions cover a slightly different thing, namely the ability to retrieve a single result by its ID. I think that I will add that functionality even though it's not spelled out in the issues.

jml replied:
Cool, thanks.

avg-I replied:
Because of my inexperience with RESTful APIs I overlooked that there were commonly accepted guidelines on designing the APIs. Now that, thank to you, I've discovered them I've updated the interface. I've implemented GET and DELETE methods and all tests are done through the HTTP API.

jml commented on benchmark/httpapi.py:
```diff
@@ -0,0 +1,215 @@
+    :ivar dict results: Stored results by their identifiers.
+    """
+    def __init__(self):
+        self.results = {}
```
Probably should make this `_results`, since it's an internal detail.

avg-I replied:
Done.

jml commented on benchmark/_interfaces.py:
```diff
@@ -0,0 +1,48 @@
+
+        :param id: The identifier of the result.
+        :return: A Deferred that fires when the result is removed.
+        """
```
Please add tests for this interface, and including tests to ensure that the two new backend classes implement this interface correctly.

avg-I replied:
This interface is not actually specified in the requirements. Perhaps I should just drop it?

jml replied:
I think the interface is a good idea, and that we should keep it. The interface tests don't have to be super thorough.

avg-I replied:
Okay.

avg-I committed:
style clean-up

avg-I committed:
switch from twisted to testtools for testing

avg-I committed:
remove completely useless DummyBackend

avg-I committed:
use constants from twisted.web.http for HTTP status codes

avg-I committed:
follow the best practices for RESTful APIs

- the "/submit" endpoint is renamed to "/benchmark-results"
- the Location header is set to point to a newly stored result
- the response code is changed to 201 Created
- the tests are updated accordingly

avg-I committed:
improve test code organization

Extract shared code into a new method, submit().
Avoid calling tests from one another, so that tested aspects are clearly
separated.

avg-I committed:
InMemoryBackend: fix retrieve and delete methods to return Deferred

avg-I committed:
continue applying best practices for RESTful interfaces

Each stored result now has an individual endpoint under
"/benchmark-results/".  GET and DELETE methods can now be used to retrieve
or remove a result correspondingly.
PUT and PATCH methods are not supported.

While there, rename the submit method of BenchmarkAPI_V1 to "post", so
that it matches the naming convention used for other methods.

test_submit_persists() is changed to use the GET HTTP request instead
of directly peeking into the backend.
Also, tests for the DELETE request are added.

avg-I committed:
install development tools (testtools for now) for travis testing

avg-I committed:
travis: use pip install -e

It turns out that "pip install .[foo]" does not work while the following
do:
- pip install pkg[foo]
- pip install -e .[foo]

So, I am going for the latter option.
See http://stackoverflow.com/questions/19096155/setuptools-and-pip-choice-of-minimal-and-complete-install/26285009#26285009

avg-I commented:
@jml I've tried to address your comments, could you please review the code again?

jml commented on benchmark/httpapi.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        request.setHeader(b'content-type', b'application/json')
+        request.setResponseCode(NO_CONTENT)
+        return self.backend.delete(id)
```
What happens if we delete something that doesn't exist?

avg-I replied:
This should be handled by `_not_found` method defined above. It converts `ResultNotFound` exception to a `404` reponse.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+            addr = listening_port.getHost()
+            self.agent = client.ProxyAgent(
+                endpoints.TCP4ClientEndpoint(
+                    reactor,
```
:question: Use `self.reactor` to use the one provided by testtools.

avg-I replied:
Done.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+        listening.addCallback(make_client)
+        endpoint = TestEndpoint(reactor, listening)
+        self.service = StreamServerEndpointService(endpoint, site)
+        self.service.startService()
```
Probably put:

```python
self.addCleanup(self.service.stopService)
```

here, and delete `tearDown`.

avg-I replied:
Done.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+
+    def test_get_deleted(self):
+        """
+        Deleted result can not be retrived.
```
Nit: retrieved.

avg-I replied:
Fixed here and at another place.

avg-I committed:
use reactor provided by testtools.TestCase instead of the global one

avg-I committed:
stop the tested service via addCleanup rather than in tearDown

avg-I committed:
spelling

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+            return _response
+
+        req.addCallback(save_response)
+        req.addCallback(client.readBody)
```
:question: Rather than doing the list trick, it's more idiomatic to do the following:

```python
req.addCallback(lambda x: client.readBody(x).addCallback(lambda y: (x, y)))
```

(giving `x` and `y` meaningful names, naturally)

And then continue to pass the extra variable through the other callbacks.

avg-I replied:
Okay. It's quite possible that using a matcher might make these techniques unnecessary.

avg-I replied:
I've re-written the code, so that it passes two values the way you suggested.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+            self.assertEqual(
+                expected_location,
+                response[0].headers.getRawHeaders(b'Location')[0]
+            )
```
:question: We could write a [matcher](http://testtools.readthedocs.org/en/latest/for-test-authors.html#matchers) for this.

avg-I replied:
I have almost no experience with testtools, so I am not sure what kind of a matcher would make most sense here.
E.g.:
* `assertThat(response, hasExpectedHeader(name, value))`
* `assertThat(response, hasExpectedLocation(location))`
* `assertThat(response, hasConsistentLocationAndID())`

Or something else?

avg-I replied:
I re-organized the code a bit and now it's less obvious (to me) that a matcher would be better than just the plain code.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+
+    def check_received_result(self, response, expected_result):
+        """
+        Response body contains the expected result.
```
Add "If it does, return the JSON decoded response body"

avg-I replied:
Done.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+            response.code, expected_code, "Incorrect response code")
+        return response
+
+    def parse_submit_response_body(self, body):
```
Please add a docstring.

avg-I replied:
Done.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+
+    def test_submit_persists(self):
+        """
+        Submitted result is stored in the backend and it can be retrived
```
Nit: retrieved.

jml commented on benchmark/test/test_httpapi.py:
```diff
@@ -0,0 +1,260 @@
+            return deleted2
+
+        req.addCallback(delete_twice)
+        return req
```
Please add tests for the following:

* POSTing a new record with the same ID as an existing record
* GETting a record that doesn't exist
* DELETEing a record that doesn't exist

avg-I replied:
Not sure if it's possible to implement the first of the suggested tests as the ID is assigned on the server side. I.e., each submitted result gets a unique ID even if it's a duplicate.

Will implement the other two.

jml commented:
Thanks Andriy. I think the roundtripping really helps clarify the API. 

A few niggles and some extra test coverage suggested. Please feel free to merge after addressing.

avg-I committed:
add / improve doc-strings

avg-I committed:
test attempts to get or delete a result that does not exist

avg-I committed:
make test_submit_response_location_header more compact

Also, the new code does not require any Python tricks.

avg-I commented:
I've addressed your comments.

Not sure if the new code in `test_submit_response_location_header` is good enough now.
But I wonder if that test is needed at all. Or, to better put it, if I need to return an ID in two places in a reponse to the POST method: the `Location` header and the response content
​It should be sufficient to set just the location and verify that the resource is accessible through it (which we already test).
